jax/_src/array.py

```diff
@@ -700,54 +700,81 @@ def make_array_from_callback(
+      the each process will call with a different ``arrays`` argument that corresponds to that processes' data.
 
   Returns:
-    A ``jax.Array`` from a sequence of ``jax.Array``\s on a single device.
+    A global ``jax.Array``, sharded as ``sharding``, with shape equal to ``shape``, and with per device
+      contents matching ``arrays``.
 
-  Example:
+  Examples:
+
+    In this single process example, we use ``make_array_from_single_device_arrays`` to create an
+    a global array.
 
     >>> import math
     >>> from jax.sharding import Mesh
     >>> from jax.sharding import PartitionSpec as P
     >>> import numpy as np
     ...
+    >>> mesh_rows = 2
+    >>> mesh_cols =  jax.device_count() // 2
+    ...
     >>> global_shape = (8, 8)
-    >>> global_mesh = Mesh(np.array(jax.devices()).reshape(2, 4), ('x', 'y'))
+    >>> global_mesh = Mesh(np.array(jax.devices()).reshape(mesh_rows, mesh_cols), ('x', 'y'))
     >>> sharding = jax.sharding.NamedSharding(global_mesh, P('x', 'y'))
     >>> inp_data = np.arange(math.prod(global_shape)).reshape(global_shape)
     ...
     >>> arrays = [
-    ...     jax.device_put(inp_data[index], d)
-    ...     for d, index in sharding.addressable_devices_indices_map(global_shape).items()]
+    ...    jax.device_put(inp_data[index], d)
+    ...        for d, index in sharding.addressable_devices_indices_map(global_shape).items()]
     ...
     >>> arr = jax.make_array_from_single_device_arrays(global_shape, sharding, arrays)
-    >>> arr.addressable_data(0).shape
-    (4, 2)
-
-    In multi-process case, if the input is process local and data parallel
-    i.e. each process receives a different part of the data, then you can use
-    `make_array_from_single_device_arrays` to create a global jax.Array
-
-    >>> local_shape = (8, 2)
-    >>> global_shape = (jax.process_count() * local_shape[0], ) + local_shape[1:]
-    >>> local_array = np.arange(math.prod(local_shape)).reshape(local_shape)
-    >>> arrays = jax.device_put(
+    >>> assert arr.shape == (8,8)
+    # arr.shape is (8,8) regardless of jax.device_count()
+
+    When using multiple processes, the most common case is the input is process local and data parallel.
+    (Equivalently, we assume each process receives a different part of the data and the
+    total number of rows is at least one per device.) In this case, the following recipe will use
+    `make_array_from_single_device_arrays` to create a global jax.Array. This recipe is very typical
+    when creating data loading pipelines.
+
+    >>> sharding = jax.sharding.NamedSharding(global_mesh, P(('x', 'y'),))
+    >>> rows_per_device = 2
+    >>> feature_length = 32
+    >>> per_host_shape = (rows_per_device * jax.local_device_count(), feature_length)
+    >>> global_shape = (jax.device_count() * rows_per_device, ) + per_device_shape[1:]
+    >>> per_host_generator = lambda : np.arange(math.prod(per_device_shape)).reshape(per_device_shape)
```

Comment:
I don't think this + below logic is right anymore? e.g. this is generating per-device, not per-device

Reply:
There was a change at one point where I wanted to do things per device but Yash pointed out a per host and per device mix was more realistic.

I don't have an amazing way to test the code but I just copy and pasted the latest version and it seemed happy? Can you help me run official read the docs?